@command_line annotation

ChrisDodd opened the pull request:
- per-target selectable support, though most should want it

fruffy commented:
Would be great to show the utility with one of the existing parametrized tests. 

https://github.com/p4lang/p4c/blob/main/backends/p4test/CMakeLists.txt#L118


I also think it makes sense to add target and arch as arguments to the annotation to make sure that only the right compiler picks up the option. This might make this parser safer.

ChrisDodd commented:
> Would be great to show the utility with one of the existing parametrized tests.
> https://github.com/p4lang/p4c/blob/main/backends/p4test/CMakeLists.txt#L118

I added a `@command_line` to these tests (and checked that the result matches), and removed the now redundant runs from CMakeLists.txt

> I also think it makes sense to add target and arch as arguments to the annotation to make sure that only the right compiler picks up the option. This might make this parser safer.

I would think it would be common to want arguments for multiple targets when you have a program that will actually compile on multiple targets.  On can use `#ifdef TARGET_FOO` around things that you only want on a specific target, though that gets messy quickly.  There's nothing terribly unsafe about these annotation, as currently they *will* flag an error for unrecognized command line options for the current target.

kfcripps commented on testdata/p4_16_samples_outputs/parser-unroll-issue3537-1-midend.p4:
```diff
@@ -34,6 +31,9 @@ parser ParserI(packet_in packet, out H hdr, inout M meta, inout standard_metadat
     state s4 {
         transition accept;
     }
+    state start {
+        transition s1;
+    }
```
Why does `start` get moved here in this test?

ChrisDodd replied:
This is something the `ParsersUroll` pass does .    I checked these outputs against the deleted ones in parser-unroll (which was the outputs for the tests with the explicit `--loopsUnroll` argument in ctest) and they match.

kfcripps replied:
Are the midend outputs changing now because:

- Previously, these outputs were the result of running the corresponding tests without `--loopsUnroll`. The outputs in p4_16_samples_outputs/parser-unroll/ were the result of running the same set of tests with `--loopsUnroll`.
- Now, we are only running each of these tests one time - only with `--loopsUnroll`. So the set of outputs corresponding to no `--loopsUnroll` were deleted and replaced by outputs of the same set of tests corresponding to having `--loopsUnroll`.

Is that correct?

ChrisDodd replied:
Yes, exactly.

ChrisDodd replied:
As far as I can see, no coverage is lost by not running these tests without --loopsUnroll as most of them are duplicates of other tests cases (control-index-hs-test*.p4 for header stack related things) that run without --loopsUnroll.  My guess is that when the ParserUnroll pass was first written, they ran all the test cases with it turned on and gathered all the things that failed in some way (by crashing in ParserUnroll probably) as these parser-unroll- tests.  Because there was no easy way to not run them multiple times, they were ok with the extra testing overhead.

Ideally we could actually do coverage testing to see what parts of the compiler are covered by tests and what tests redundantly cover the same things, so we could prune the test suite better.

Of the 16 tests covered here:
12 of the 16 don't actually do any unrolling, so the coverage by that test without --loopsUnroll is a subset of the coverage with --loopsUnroll.
Those that do unroll involve header_stacks (since a header_stack is pretty much required to have a parser that can be unrolled in the first case), and similar cases are covered elsewhere (the control-index-hs tests and many others).

However, we could remove the second change from this PR (which has the test changes in order demonstrate how `@command_line` works) if that is too much of a worry.

ChrisDodd replied:
As for who wrote these tests -- the issue3537 and issue4006 (3 cases) were checked in by @kfcripps and the uninitialized-slice (2 cases) were checked in by @asl.  The rest were all by @VolodymyrPeschanenkoIntel when the ParserUnroll pass was first checked in.

kfcripps reviewed (commented):
I doubt that AMD's backend would want this (I think @asl mentioned reasons that it might not be a good idea (although I don't personally have any opinion about whether it's a good or bad idea)), but as long as it's easy for a backend to toggle this annotation to off (as you've done in this PR), then I'm not opposed to these changes.

This reminds me of issue https://github.com/p4lang/p4c/issues/4907. When I created it, I was thinking of having a way to specify compilation options in the source file some other way, for example via the LLVM `Lit` (https://llvm.org/docs/CommandGuide/lit.html) utility.

@sepanchi Sergey, have you had a chance to give any thought to this?

Maybe adding this annotation will avoid the need to use something like `Lit`.

vlstill commented on frontends/common/applyOptionsPragmas.h:
```diff
@@ -76,9 +76,13 @@ class ApplyOptionsPragmas : public Inspector {
  public:
     std::optional<CommandLineOptions> tryToParse(const IR::Annotation *annotation) override;
 
+    explicit P4COptionPragmaParser(bool sCLP) : supportCommandLinePragma(sCLP) {}
```
nitpick: I would just spell fully the long name for the sake of editors that use the signature to help the programmer with calling the constructor.

asl reviewed (approved):
I'm still not being a big fan of mixing two orthogonal things – command line options of a particular implementation and overall semantics of enable/disable several things.

But given that it is an opt-in feature in the annotation parser, I guess it should be ok.
